{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1 answer:\n",
    "Even though this answer's code could produce a shorter output, I opted for this extended version because it serves as a check that this script is doing what it's supposed to be doing.\n",
    "\n",
    "I used the csv library to handle the input files (\"LocusGene.tsv\" and \"Germplasm.tsv\") because it is very useful and how easy it is to use. \n",
    "\n",
    "The first two variables I declared (\"tmp\" and \"mismatch\") are used to create the output. The former serves as a counter of the line that is being checked in each iteration of the loops and the latter serves as a counter of the times two lines are not the same in both documents. \n",
    "\n",
    "Next, I opened the files previously mentioned. As adding information to them isn't necessary, I opened them in read mode (the *\"r\"* part of the open call makes the files open in this mode). Note that I didn't close the connection to the files yet. This is because I still needed to make more operations on them in another problem's answer. \n",
    "Then, using the csv.DictReader function I loaded the data from both files to their respective functions (\"locus_data\" and \"germplasm_data\". Since these are \".tsv\" files, the delimiter is set to a tab (\"\\t\"). Even though there are no quotation marks in the documents, I set it to double quotation marks (\") just in case in the future I wanted to use them. \n",
    "\n",
    "As the way the nested loops work is quite counter-intuitive, I will explain it in detail:\n",
    "Normally, when a for loop is nested into another, the resulting behaviour is that for each iteration of the first loop, the entirety of the nested loop is ran. In this particular situation, that means that the nested loops would cause the first line of the \"LocusGene.tsv\" file to be compared to **every** single line in the \"Germplasm.tsv\" document. But as the output shows, that's not what is happening. There are two reasons for this:\n",
    "* After all the code in the nested loop, I inserted a *\"break\"*. This means that instead of running this loop completely, only one iteration is actually run. That is because \"break\" causes the current loop to be terminated, and resumes execution at the next statement, which in this case is the next iteration of the first loop (the one that is not nested). \n",
    "* Ok, so for now the loops have compared the first line in the \"LocusGene.tsv\" and \"Germplasm.tsv\" files and the nested loop has been terminated, so now what should happen is that the second line of the LocusGene file should be compared with the **first** line on the Germplasm file again. But again, the output shows something different: the second line of both files is being compared. This is because I am reading from a file! When the for loop reads a line from either file, said file's pointer is set to the end of that line. This means that after reading one line, that line is no longer \"visible\" to the DictReader. The result of this interaction is that after comparing the fist line of both documents, and terminating the nested loop, the first loop reads the following line, which is the second one because of where the pointer is! Then, the second loop reads the \"first\" line in the Germplasm file, but since the pointer is at the end of the first line, the second line is what is really read. \n",
    "\n",
    "This is pretty hard to explain, so I will try to summarize:\n",
    "The for loop nested structure in my code works because when the files are read, the file pointer is set to the end of the last line read. With each iteration of the loops, the next lines are read and the pointer is subsecuentially placed on the end of those lines. Also, since I introduced a \"break\" in the nested loop, only one iteration is run, resulting in the designed behaviour consisting of comparing the first line in the LocusGene file with the first line in the Germplasm file, the second with the second and so on. \n",
    "Just to prove my explanation, I created a copy of this code in another code cell, but I commented out the \"break\" part. Check that explanation please.\n",
    "\n",
    "As for what the loops do themselves, they simply print a string stating the line number that is being compared, show the comparison *per se* and showing the result of the comparisons thanks to the if/else block. When I say that the lines from both documents are being compared, I mean that the \"Locus\" column from both documents is being compared. This is done in the *\" if row1[\"Locus\"] == row2[\"Locus\"]\"* line. That is because the DictReader function outputs a variant of a dictionary created from separating each column of the document. Its keys are the header line's text for each column (if present, otherwise they can be declared in the DictReader statement) and the values are, well, the values of each row corresponding to each column (in this case, for example, the first element of the dictionary would be {\"Locus\" : \"AT1G01040\"}, the second {\"Gene\" : \"DCL1\"} and so on until the first line is recorded. Then the loop would iterate over the second line and fill in the values in that row). If the lines are not the same, the \"mismatch\" counter is incremented. At the end of each iteration the \"tmp\" counter increments as well to show the next lines to be checked. The \"tmp\" function is set to 1 at the beginning because even though technically the first lines to be compared are number 0 (at least in Python), it is easier for a human to understand if the count begins in 1. \n",
    "\n",
    "Finally, the final results are created. If the \"mismatch\" variable's value is 0, that means that all comparisons showed that the lines were the same, so both documents are in the same order. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#First of all, I have to import the csv library to be able to use its functions.\n",
    "import csv\n",
    "\n",
    "#These variables are used in creating the output of the file.\n",
    "tmp = 1\n",
    "mismatch = 0\n",
    "\n",
    "#Next, I open both files in read mode and read their contents with the csv.DictReader function\n",
    "locus_file = open(\"LocusGene.tsv\", \"r\")\n",
    "locus_data = csv.DictReader(locus_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "germplasm_file =  open (\"Germplasm.tsv\", \"r\")\n",
    "germplasm_data = csv.DictReader(germplasm_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "\n",
    "#The nested for loops allow me to easily compare both files.\n",
    "for row1 in locus_data:   \n",
    "    for row2 in germplasm_data:\n",
    "        print(\"Now checking line number: {}\".format(tmp))\n",
    "        print(row1[\"Locus\"] + \" vs. \" + row2[\"Locus\"])\n",
    "        if row1[\"Locus\"] == row2[\"Locus\"]:\n",
    "            print(\"Result: both are the same\\n\")\n",
    "        else:\n",
    "            print(\"Result: both are not the same\\n\")\n",
    "            mismatch += 1\n",
    "        tmp +=1    \n",
    "        break\n",
    "        \n",
    "#The final results are created here.        \n",
    "print (\"FINAL RESULTS:\")                \n",
    "if mismatch == 0:\n",
    "    print(\"All lines are in the same order!\")\n",
    "else:\n",
    "    print(\"Warning! {} lines are not in the same order\".format(mismatch))   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Explanation of the same code but with the break disabled.\n",
    "As we can see in the output, when the \"break\" is not present on the nested loop the behaviour is as explained previously: for each iteration of the first loop all the nested loop is run. This means that the first line of the LocusGene document is being compared to all the lines in the Germplasm document (as we can see in the fact that the first element of the comparison is always \"AT1G0140\").\n",
    "Furthermore, my point about the pointer being set to the end of each line when it is read is reflected in the fact that **only one iteration of the first loop is showing output**. This is because after the first iteration of the first loop the whole Germplasm document has been read, so the file pointer is at the end of said file! This means that all the other iterations of the first loop are comparing the LocusGene to a functionally *empty* document, so the nested loop can't run any more iterations, causing the lack of output. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First of all, I have to import the csv library to be able to use its functions.\n",
    "import csv\n",
    "\n",
    "#These variables are used in creating the output of the file.\n",
    "tmp = 1\n",
    "mismatch = 0\n",
    "\n",
    "#Next, I open both files in read mode and read their contents with the csv.DictReader function\n",
    "locus_file = open(\"LocusGene.tsv\", \"r\")\n",
    "locus_data = csv.DictReader(locus_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "germplasm_file =  open (\"Germplasm.tsv\", \"r\")\n",
    "germplasm_data = csv.DictReader(germplasm_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "\n",
    "#The nested for loops allow me to easily compare both files.\n",
    "for row1 in locus_data:   \n",
    "    for row2 in germplasm_data:\n",
    "        print(\"Now checking line number: {}\".format(tmp))\n",
    "        print(row1[\"Locus\"] + \" vs. \" + row2[\"Locus\"])\n",
    "        if row1[\"Locus\"] == row2[\"Locus\"]:\n",
    "            print(\"Result: both are the same\\n\")\n",
    "        else:\n",
    "            print(\"Result: both are not the same\\n\")\n",
    "            mismatch += 1\n",
    "        tmp +=1    \n",
    "        #break\n",
    "        \n",
    "#The final results are created here.        \n",
    "print (\"FINAL RESULTS:\")                \n",
    "if mismatch == 0:\n",
    "    print(\"All lines are in the same order!\")\n",
    "else:\n",
    "    print(\"Warning! {} lines are not in the same order\".format(mismatch))   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 2 answer:\n",
    "Since sql magic is easier to use, its output is much easier to read and the following commands are basic, I chose to use it.\n",
    "\n",
    "First, I had to run the code that is used to connect to MySQL in Docker. \n",
    "\n",
    "Then I had to create a database for this exam, called \"exam2\", and inside of said database I created two tables. Since the \"Locus\" column is shared between both input files and locus' codes shouldn't appear more than once, I decided to use them as the primary keys for their respective tables. \n",
    "The locus column is the primary key, which means that its values **must** be unique in the entire table. It also means that it is the column that will be used to link both tables in this case. The values contained in this column must be VARCHAR, variable characters up to length 15 (even though all codes in the input file are 9 characters long, I just wanted to make sure that if there is any AGI code that is longer it can fit). The \"protein_length\" column is to be filled with INTEGERs. I made it so all columns must contain information (basically, they can't be null).\n",
    "The germplasm table has different columns but a similar configuration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Commands to connect to MySQL in Docker. \n",
    "%load_ext sql\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now, I create a database for this exam (called \"exam2\")\n",
    "%sql CREATE DATABASE exam2;\n",
    "%sql USE exam2; #and I tell sql that I want to work with said database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%sql DROP TABLE locus_gene; #This is here just in case I need to delete the table and start over.\n",
    "\n",
    "#This creates the \"locus_gene\" table.\n",
    "%sql CREATE TABLE locus_gene(locus VARCHAR(15) PRIMARY KEY NOT NULL, \\\n",
    "                             gene VARCHAR(20) NOT NULL, protein_length INTEGER NOT NULL);\n",
    "#This is used to check that the table was created correctly.\n",
    "%sql DESCRIBE locus_gene; "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#%sql DROP TABLE germplasm;\n",
    "%sql CREATE TABLE germplasm(locus VARCHAR(15) PRIMARY KEY NOT NULL, \\\n",
    "                            germplasm VARCHAR(25) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 3 answer:\n",
    "As I have previously explained, right now the pointer in both files is at their end. Before doing anything else with them, I have to reset the pointers' position to the beginning of the first data line. I did this by setting them at the beginning of the file and then reading the first line, the header line, skipping it and setting the pointers in the desired position. \n",
    "\n",
    "Once again I have to iterate over both files. I did this with a for loop for each one. Both of them are essentially the same. As I said before, \"row\" is, in this case, a variation of dictionary that contains the name of each column as keys and the values corresponding to each entry for those columns as values. \n",
    "In the SQL commands I'm inserting into each of the database's columns their corresponding values by accessing the values in the row dictionary and adding them in the correct place. The .format method allows me to introduce \"placeholders\" in the SQL command ({}) and then passing the corresponding Python variables to each placeholder in order of appearance. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#First of all, I have to import the pymysql.cursors library to be able to use its functions.\n",
    "import pymysql.cursors\n",
    "\n",
    "#Set the pointer to the beginning of the first line of data (skipping the header line)\n",
    "locus_file.seek(0)\n",
    "locus_file.readline()\n",
    "germplasm_file.seek(0)\n",
    "germplasm_file.readline()\n",
    "\n",
    "#stablish a connection with the database.\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='exam2',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "#insert the data\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for row in locus_data:\n",
    "            sql = \"\"\"INSERT INTO locus_gene (locus, gene, protein_length)\n",
    "            VALUES ('{}', '{}', {});\"\"\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "            cursor.execute(sql) \n",
    "        for row in germplasm_data:\n",
    "            sql = \"\"\"INSERT INTO germplasm (locus, germplasm, phenotype, pubmed)\n",
    "            VALUES ('{}', '{}', '{}', {});\"\"\".format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "            cursor.execute(sql) \n",
    "    connection.commit() #this commits the changes to the database.\n",
    "finally:\n",
    "    connection.close() #Here I'm closing the connection to the database.\n",
    "    locus_file.close() #Here I'm closing the connection to the \"LocusGene.tsv\" file.\n",
    "    germplasm_file.close() #Here I'm closing the connection to the \"Germplasm.tsv\" file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Here I'm checking if the data was inserted correctly.\n",
    "%sql SELECT * FROM germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 4 answer:\n",
    "As I was reading the things that were required for problem 4 and the note at the end reminding us about the DRY rule I thought to myself: why not try to make a multi-purpose report-making function? I could even use it in the future to make reports of different kinds and shapes. The problem is that this weekend I didn't have a lot of time to develop the function much more than the bare minimum to be able to do the things this problem asks for. I didn't have time to make the function user-proof (making it not crash if you don't pass the correct data types as arguments, making it completely abstract and other things (for example, I hard coded some of the table names in the parts of the function that count the amount of entries and the average protein length in each chromosome because there was no immediate need to make those parts completely customizable just yet). If I had had enough time, I would have made a lot of improvements, but right now is a barebones version. \n",
    "\n",
    "First of all, I defined the function \"ReportMaker\". I won't explain what each argument does since the docstring already has plenty of information about the functioning of the function and each of its arguments. \n",
    "The docstring is the really big multi line text that is just after the def ReportMaker line. Its purpose is to explain the user what the function and its arguments do. The easiest way to check this information is by executing \"help(ReportMaker)\".\n",
    "\n",
    "Then, the function tries to connect to the selected database. As for the purposes of this course the connection block is always the same and the only part that needs to be modified is the \"db\" variable, I created a try/except block. The way this block works is the following: \n",
    "* First, the try clause is executed.\n",
    "* If there is no exception during that execution, the except clause is skipped.\n",
    "* If an exception occurs during that execution, the rest of the try clause is skipped and if the type of the exception matches the one named after \"except\", the except clause is executed, prompting the user to check if they introduced the database name correctly.\n",
    "\n",
    "After that, there is a block of conditionals that handle the different modes of the function:\n",
    "* \"individually\" block: **This block is NOT used for this exam's answers**, but it is a simple example that works similarly to the rest of the more complex modes. First of all I used the str.lower() method to make this if statement be case insensitive to check if the mode is \"individually\". The way it works is that it makes the string assignated to the \"mode\" variable passed by the user lowercase before checking if it is \"individually\". Then, if there was no \"gene\" argument in the call to the function, it displays the selected table. Otherwise, it filters the selected table so that it only shows the entry in which the \"gene\" column matches the name inserted. The only newly introduced concept here is the filtering, which is made with the \"LIKE\" part of the sql query. The LIKE operator is used in a WHERE clause to search for a specified pattern in a column. **Note: notice that some of the sql placeholders ({}) have quotation marks (') and others don't.** It is better explained with an example. In this case, the query \"SELECT * FROM {} WHERE gene LIKE '{}'\".format(table1, gene) would look like this in SLQ itself: SELECT * FROM table1 WHERE gene LIKE \"gene\". Since the pattern that LIKE searches for must be quoted, its placeholder must be quoted as well.\n",
    "\n",
    "* \"joined\" block: The same check for the value of \"gene\" that happens in the previous block happens in this block. If there is no \"gene\" to filter by, the function outputs the two selected tables joined by their two respective selected columns. This is possible thanks to the \"WHERE\" clause. In this case it would be: \"WHERE table1.join_col1 = table2.join_col2\", effectively joining both tables. If there is a \"gene\" value, it is used to filter, just the same as before. The only new operator is \"AND\", that makes the SQL query check for various conditions. \n",
    "\n",
    "* \"count\" block: Since all of the results obtained from this mode are made so from a loop, I had to create the list \"results\" to be able to append the results to it in each iteration of the loops. In the \"by_chromosome\" count type, the for loop iterates in the range(1, chromosome_count + 1). That is because the chromosomes of all species (at least all I know of) start by the number 1, and since Python's first index is 0 I had to specify the range to begin by 1. Then, the chromosome count of the species is used as the last part of the range, but I had to add 1 to it because Python's ranges don't include the last number included in them. I could have hard coded the chromosome count to be 5, since this exam's data is about *Arabidopsis*, and it only has 5 chromosomes- But abstracting this part of the problem was easy enough so I decided to go for it. The only new parts are the \"COUNT\" and \"AVG\" operators, as well as a new filtering pattern;\n",
    "    - \"COUNT\": it literally counts entries that fulfill certain conditions. Also, if an \"AS\" operator is introduced in its statement, the number of entries that fulfill the condition(s) will be named as the string following AS. \n",
    "    - \"AVG\": it functions the same as the \"COUNT\" operator, but this one averages values.\n",
    "    - Filtering pattern: The new filtering pattern is \"ATnG%\". The \"%\" represents zero, one, or multiple characters and the n is the chromosome number. So for chromosome 2 this pattern would mean: look for entries that begin with AT2G.\n",
    "    \n",
    "In the main code, I called the function with the necessary parameters to create the required output for this problem. \n",
    "Note that, since the function returns the \"results\" variable, it is now available in the global scope and can be used outside of the function itself!\n",
    "\n",
    "Since I'm planning to create a \".tsv\" file, **I strongly recommend opening the output document in LibreOffice calc or similar**, as it is MUCH easier to read the output in those programs.\n",
    "\n",
    "Since the \"results\" variable is a list of dictionaries, and it is stored in the \"result\" variable as dictionaries in the loop, I used the dict.keys() method to extract the key values from it. Those keys correspond to the column names in the SQL database, so I used them as the header of the first and second reports. Each column name is separated by a tab (\"\\t\") because I want the ouput to be a \".tsv\" file.\n",
    "Then, to fill the report with the data from each row I simply accessed the values associated with each key and inserted them in their corresponding output file column with a loop.\n",
    "\n",
    "The only differences appear in the reports 3 and 4, where I had to append the resulting dictionaries for each query into a list. Because of that, the \"results\" variable in these cases is a list of lists of dictionaries (a list containing lists containing dicionaries) and the \"result\" variable in the loops is therefore a list of dictionaries, and had to be accessed as one. \n",
    "This means that \"tmp = result[0][\"C_{}\".format(n)]\" basically translates to: in the result variable, look for the first element of the list of dictionaries, and inside look for the \"C_n\" key (where n is the chromosome number that is being checked), and assing that value to the tmp variable.\n",
    "I decided to add a *control* to the 3rd report, so I added a line with the total number of entries, which is the value of tmp2, that is the result of adding tmp to tmp2 in each of the loop's iterations. Tmp is the number of entries in a given chromosome.\n",
    "\n",
    "Finally, I open the ouput file, called \"report_file.tsv\". This is done in a with statement so that I don't forget to close the connection to the file after using it. \n",
    "Even though I could have created the file and accessed it in append mode, I thought it would be easier if the script created the file each time it is run instead of appending."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def ReportMaker (database, table1, mode = \"individually\", table2 = None,\n",
    "                 join_col1 = None, join_col2 = None, gene = None, count_type = None, chromosome_count = None):\n",
    "    \"\"\"ReportMaker creates highly customizable reports.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    database:\n",
    "        string --The name of the database you want to create reports from. --OBLIGATORY\n",
    "    \n",
    "    table1: \n",
    "        string --The name of the table you want to appear in the report. --OBLIGATORY\n",
    "    \n",
    "    mode: \n",
    "        string --Currently, the available modes are:                           --OBLIGATORY\n",
    "                 *\"individually\": the report will contain only one table. \n",
    "                 *\"joined\": the report will TRY to join two tables with a common column.\n",
    "                 *\"count\": the report will show you a table and count as selected in the count_type parameter.\n",
    "                 \n",
    "    table2:\n",
    "        string --If you selected the \"joined\" mode, the name of the second table. OBLIGATORY if you use \n",
    "                 mode \"joined\"--OPTIONAL\n",
    "                 \n",
    "    join_col1:\n",
    "        string --The name of the column FROM THE FIRST TABLE you want to use to join the tables. OBLIGATORY if you use\n",
    "                 mode \"joined\". --OPTIONAL\n",
    "                 \n",
    "    join_col2:\n",
    "        string --The name of the column FROM THE SECOND TABLE you want to use to join the tables. OBLIGATORY if you use\n",
    "                 mode \"joined\". --OPTIONAL\n",
    "    \n",
    "    gene:\n",
    "        string --You can filter the report with the names of one gene. --OPTIONAL\n",
    "        \n",
    "    count_type:\n",
    "        string --table1 will be used to count. Currently, the available count modes are:\n",
    "                 *\"by_chromosome\": counts the number of entries for each chromosome.\n",
    "                 *\"avg_prot_len\": the report contains the average protein length for the genes on each chromosome.\n",
    "                 OBLIGATORY if you use mode \"count\"--OPTIONAL\n",
    "    chromosome_count:\n",
    "        integer --Number of chromosomes the species the table is about has. \n",
    "                  OBLIGATORY if you use mode \"count\". --OPTIONAL\n",
    "    \"\"\" \n",
    "    import pymysql.cursors\n",
    "    \n",
    "    try:\n",
    "        connection = pymysql.connect(host='localhost',\n",
    "                                 user='root',\n",
    "                                 password='root',\n",
    "                                 db= database,\n",
    "                                 charset='utf8mb4',  \n",
    "                                 cursorclass=pymysql.cursors.DictCursor)\n",
    "    \n",
    "    #this except block promts the user to ckeck if their MySQL server is running as this is the exception that occurs \n",
    "    #when it isn't\n",
    "    except OperationalError:\n",
    "        print(\"\"\"Something went wrong! Check that your MySQL server is running\"\"\")\n",
    "        \n",
    "    #this except block prompts the user to check if they introduced the database name correctly, \n",
    "    #as every other parameter in the connection statement should work (in this course's environment)\n",
    "    except InternalError:\n",
    "        print(\"Something went wrong! Check the name you passed as the 'database' argument.\")\n",
    "        \n",
    "    with connection.cursor() as cursor: \n",
    "        #This block of code handles the \"individually\" mode of the function.\n",
    "        if mode.lower() == \"individually\":\n",
    "            #if there is no \"gene\" argument passed, then the function simply shows the whole selected table.\n",
    "            if gene == None:\n",
    "                sql = \"SELECT * FROM {}\".format(table1)\n",
    "                cursor.execute(sql) \n",
    "                results = cursor.fetchall()\n",
    "            #otherwise, the value of \"gene\" will be used to filter the table.    \n",
    "            else:\n",
    "                sql = \"SELECT * FROM {} WHERE gene LIKE '{}'\".format(table1, gene) \n",
    "                cursor.execute(sql)\n",
    "                results = cursor.fetchall()\n",
    "                \n",
    "        #This block of code handles the \"joined\" mode of the function\n",
    "        elif mode.lower() == \"joined\":\n",
    "            #if \"gene\" is None, then the function shows the two selected tables joined by the two selected columns.\n",
    "            if gene == None:\n",
    "                sql = \"SELECT * FROM {}, {} WHERE {}.{}= {}.{}\".format(table1, table2, table1,\n",
    "                                                                       join_col1, table2, join_col2)\n",
    "                cursor.execute(sql) \n",
    "                results = cursor.fetchall()\n",
    "            #otherwise, the value of gene will be used to filter the table.    \n",
    "            else:\n",
    "                sql = \"SELECT * FROM {}, {} WHERE {}.{}= {}.{} AND gene LIKE '{}'\".format(table1, table2, table1, \n",
    "                                                                                          join_col1, table2,\n",
    "                                                                                          join_col2, gene) #mirar por que en una hace falta las comillas y en otra no\n",
    "                cursor.execute(sql)\n",
    "                results = cursor.fetchall()\n",
    "                \n",
    "        #Finally, this block of code handles the \"count\" mode of the function.    \n",
    "        elif mode.lower() == \"count\":\n",
    "            #this list will be used to store all the results from each iteration of the loops.\n",
    "            results = []\n",
    "            #this part of the code counts the amount of entries for each chromosome.\n",
    "            if count_type.lower() == \"by_chromosome\":\n",
    "                for n in range(1, chromosome_count + 1):\n",
    "                    sql = \"SELECT COUNT(locus) AS 'C_{}' FROM {} WHERE locus LIKE 'AT{}G%'\".format(n, table1, n)\n",
    "                    cursor.execute(sql)\n",
    "                    results.append(cursor.fetchall())\n",
    "                    \n",
    "            #this part of the code averages the protein length for each chromosome.        \n",
    "            elif count_type.lower() == \"avg_prot_len\":\n",
    "                for n in range(1, chromosome_count + 1):\n",
    "                    sql = \"SELECT AVG(protein_length) AS 'AVG_{}' FROM locus_gene WHERE locus LIKE 'AT{}G%'\".format(n, n)\n",
    "                    cursor.execute(sql)\n",
    "                    results.append(cursor.fetchall())\n",
    "                    \n",
    "            #If the \"count_type\" value passed in the call to the function isn't one of the two currently available,\n",
    "            #I raise an exception that prompts the user to check the help for the function.\n",
    "            else:\n",
    "                raise InternalError(\"\"\"The count type you selected wasn't recognized by the function.\n",
    "                                            Check 'help(ReportMaker)' for a detailed list of count types'\"\"\")\n",
    "        #If the \"mode\" value selected isn't one of the currently available, I raise an exception that prompts\n",
    "        #the user to check the help command for this function.\n",
    "        else:\n",
    "            raise InternalError(\"\"\"The mode you selected wasn't recognized by the function. \n",
    "                                  Check 'help(ReportMaker)' for a detailed list of modes.\"\"\")\n",
    "\n",
    "    #the \"results\" variable is returned from the function so it is available in the global scope.\n",
    "    return results\n",
    "\n",
    "#------------------\n",
    "#   Main code\n",
    "#------------------\n",
    "\n",
    "#help(ReportMaker) #This is here just in case someone wants to see how it looks.\n",
    "\n",
    "#This part of the code handles the creation of the text output for report 1.\n",
    "result1 = ReportMaker(database =\"exam2\", table1 = \"locus_gene\", table2 = \"germplasm\", mode = \"joined\", \n",
    "                     join_col1 = \"locus\", join_col2 = \"locus\")\n",
    "\n",
    "out_file1_header = \"\"\n",
    "for key in result1[0].keys():\n",
    "    out_file1_header += key + \"\\t\"\n",
    "out_file1_header += \"\\n\"\n",
    "\n",
    "out_file1_body = \"\"\n",
    "for result in result1:\n",
    "    out_file1_body  += result[\"locus\"] + \"\\t\" + result[\"gene\"] + \"\\t\" + str(result[\"protein_length\"]) + \"\\t\" \\\n",
    "    + result[\"germplasm.locus\"] + \"\\t\" + result[\"germplasm\"] + \"\\t\" + result[\"phenotype\"] + \"\\t\" \\\n",
    "    + str(result[\"pubmed\"]) + \"\\n\"\n",
    "\n",
    "#This part of the code handles the creation of the text output for report 2.\n",
    "result2 = ReportMaker(database =\"exam2\", table1 = \"locus_gene\", table2 = \"germplasm\", mode = \"joined\", \n",
    "                     join_col1 = \"locus\", join_col2 = \"locus\", gene = \"SKOR\")\n",
    "\n",
    "result3 = ReportMaker(database =\"exam2\", table1 = \"locus_gene\", table2 = \"germplasm\", mode = \"joined\", \n",
    "                     join_col1 = \"locus\", join_col2 = \"locus\", gene = \"MAA3\")\n",
    "\n",
    "out_file2 = \"\"\n",
    "for result in result2:\n",
    "    out_file2 += result[\"locus\"] + \"\\t\" + result[\"gene\"] + \"\\t\" \\\n",
    "    + str(result[\"protein_length\"]) + \"\\t\" + result[\"germplasm.locus\"] + \"\\t\" \\\n",
    "    + result[\"germplasm\"] + \"\\t\" + result[\"phenotype\"] + \"\\t\" + str(result[\"pubmed\"]) + \"\\n\"\n",
    "    \n",
    "for result in result3:\n",
    "    out_file2 += result[\"locus\"] + \"\\t\" + result[\"gene\"] + \"\\t\" + str(result[\"protein_length\"]) + \"\\t\" \\\n",
    "    + result[\"germplasm.locus\"] + \"\\t\" + result[\"germplasm\"] + \"\\t\" + result[\"phenotype\"] + \"\\t\" \\\n",
    "    + str(result[\"pubmed\"]) + \"\\n\"    \n",
    "\n",
    "result4 = ReportMaker(database = \"exam2\", table1 = \"locus_gene\", mode = \"count\", count_type = \"by_chromosome\",\n",
    "                      chromosome_count = 5)\n",
    "\n",
    "#This part of the code handles the creation of the text output for report 3.\n",
    "out_file3 = \"\"\n",
    "n = 1\n",
    "tmp2 = 0\n",
    "for result in result4:\n",
    "    tmp = result[0][\"C_{}\".format(n)]\n",
    "    out_file3 += \"Number of entries for chromosome {}:\".format(n) + \"\\t\" + str(tmp) + \"\\n\"\n",
    "    tmp2 += tmp\n",
    "    n += 1\n",
    "out_file3 += \"Total number of entries:\" + \"\\t\" + str(tmp2) + \"\\n\"\n",
    "\n",
    "result5 = ReportMaker(database = \"exam2\", table1 = \"locus_gene\", mode = \"count\", count_type = \"avg_prot_len\",\n",
    "                      chromosome_count = 5)\n",
    "\n",
    "#This part of the code handles the creation of the text output for report 4.\n",
    "out_file4 = \"\"\n",
    "m = 1\n",
    "for result in result5:\n",
    "    tmp = result[0][\"AVG_{}\".format(m)]\n",
    "    out_file4 += \"Average protein length for chromosome {}\".format(m) + \"\\t\" + str(tmp) + \"\\n\"\n",
    "    m += 1\n",
    "\n",
    "#This part of the code writes every previously created output in the output file.    \n",
    "with open (\"reports_file.tsv\", \"w\") as out_file:\n",
    "    out_file.write(\"Report 1: Joined tables \\n\")\n",
    "    out_file.write(out_file1_header)\n",
    "    out_file.write(out_file1_body)\n",
    "    out_file.write(\"\"\"Comments on Report 1: As we can see, both files were in the same order. The first three columns\n",
    "    correspond to the 'LocusGene.tsv' file and the last four correspond to the file 'Germplasm.tsv'\\n\"\"\")\n",
    "    out_file.write(\"\\nReport 2: Joined tables filtered by Genes 'SKOR' and 'MAA3' \\n\")\n",
    "    out_file.write(out_file1_header)\n",
    "    out_file.write(out_file2)\n",
    "    out_file.write(\"\\nReport 3: Number of entries for each chromosome\\n\")\n",
    "    out_file.write(out_file3)\n",
    "    out_file.write(\"\"\"Comments on Report 3: As we can see, chromosome 1 has 4 entries, \n",
    "    chromosome 2 has 4, chromosome 3 has 9, chromosome 4 has 8 and chromosome 5 has 7 entries.\n",
    "    That makes a total of 32 entries, the total amount that was found in the input files\\n\"\"\")\n",
    "    out_file.write(\"\\nReport 4: Average protein length for each chromosome \\n\")\n",
    "    out_file.write(out_file4)\n",
    "    out_file.write(\"\"\"Comments on Report 4: As we can see, chromosome 1 has an average protein length of 258.75, \n",
    "    chromosome 2 has an average protein length of 215.5, chromosome 3 has an average protein length of 252, \n",
    "    chromosome 3 has an average protein length of 277.5 and finally chromosome 5 has an average protein length \n",
    "    of 271.2857 \"\"\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
