{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Solutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# First, connect to the mysql server with sqlmagic\n",
    "\n",
    "%load_ext sql\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n",
      "good\n"
     ]
    }
   ],
   "source": [
    "file1 = open('./Germplasm.tsv', 'r')\n",
    "file2 = open('./LocusGene.tsv', 'r')\n",
    "\n",
    "#discard headers\n",
    "disc = file1.readline()\n",
    "disc = file2.readline()\n",
    "\n",
    "for line_file1 in file1.readlines():\n",
    "    line_file2 = file2.readline()\n",
    "    line_file1.strip()\n",
    "    line_file2.strip()\n",
    "    (loc,germ,phen,pm) = line_file1.split(\"\\t\") # split on tab\n",
    "    (loc2,gene,prot) = line_file2.split(\"\\t\") # split on tab\n",
    "    if loc == loc2:  # compare content of first column\n",
    "        print(\"good\")\n",
    "    else:\n",
    "        print(\"NOT ALIGNED\")\n",
    "\n",
    "file1.close\n",
    "file2.close\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>examanswer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('examanswer',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',)]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "%sql show databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Tables_in_examanswer</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('genes',), ('germplasm',)]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql drop database examanswer\n",
    "%sql create database examanswer\n",
    "%sql use examanswer\n",
    "\n",
    "# Data Model:  germplasm table has an ID.  Gene table has a \"germplasm_id\" field, pointing to germplasm's  \"id\" field\n",
    "# Locus\tgermplasm\tphenotype\tpubmed\n",
    "%sql create table germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(10), germplasm VARCHAR(30), phenotype VARCHAR(10000), pubmed INTEGER)\n",
    "# Locus\tGene\tProteinLength\n",
    "%sql create table genes(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, germplasm_id INTEGER, gene VARCHAR(10), protlength INTEGER)\n",
    "%sql show tables\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AT1G01040 CS3828 Increased abundance of miRNA precursors. 17369351\n",
      "\n",
      "AT1G01040 DCL1 332\n",
      "\n",
      "[{'last_insert_id()': 1}]\n",
      "1\n",
      "AT1G01060 lhy-101 The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant. 16891401\n",
      "\n",
      "AT1G01060 LHY 290\n",
      "\n",
      "[{'last_insert_id()': 2}]\n",
      "2\n",
      "AT1G01140 SALK_058629 hypersensitive to low potassium media 17486125\n",
      "\n",
      "AT1G01140 CIPK9 223\n",
      "\n",
      "[{'last_insert_id()': 3}]\n",
      "3\n",
      "AT1G01220 SALK_012400C fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected. 18199744\n",
      "\n",
      "AT1G01220 FKGP 190\n",
      "\n",
      "[{'last_insert_id()': 4}]\n",
      "4\n",
      "AT2G03720 SALK_042433 Multiple straight hairs 16367956\n",
      "\n",
      "AT2G03720 MRH6 189\n",
      "\n",
      "[{'last_insert_id()': 5}]\n",
      "5\n",
      "AT2G03800 gek1-1 Ethanol hypersensitivity. 15215505\n",
      "\n",
      "AT2G03800 GEK1 196\n",
      "\n",
      "[{'last_insert_id()': 6}]\n",
      "6\n",
      "AT2G04240 xerico Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype. 17933900\n",
      "\n",
      "AT2G04240 XERICO 256\n",
      "\n",
      "[{'last_insert_id()': 7}]\n",
      "7\n",
      "AT2G05210 pot1-1 No visible phenotype. 17627276\n",
      "\n",
      "AT2G05210 POT1A 221\n",
      "\n",
      "[{'last_insert_id()': 8}]\n",
      "8\n",
      "AT3G02130 rpk2-2 The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants. 17419837\n",
      "\n",
      "AT3G02130 RPK2 284\n",
      "\n",
      "[{'last_insert_id()': 9}]\n",
      "9\n",
      "AT3G02140 afp4-1 Decreased germination on high concentrations of glucose and sorbitol. 18484180\n",
      "\n",
      "AT3G02140 TMAC2 300\n",
      "\n",
      "[{'last_insert_id()': 10}]\n",
      "10\n",
      "AT3G02230 rgp1-1 rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls. 21478444\n",
      "\n",
      "AT3G02230 RGP1 301\n",
      "\n",
      "[{'last_insert_id()': 11}]\n",
      "11\n",
      "AT3G02260 tir3-1 RGLG1:rglg1 rglg2 The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence. 17586653\n",
      "\n",
      "AT3G02260 BIG 279\n",
      "\n",
      "[{'last_insert_id()': 12}]\n",
      "12\n",
      "AT3G02310 sep2-1 Non-described subtle phenotype. 10821278\n",
      "\n",
      "AT3G02310 SEP2 175\n",
      "\n",
      "[{'last_insert_id()': 13}]\n",
      "13\n",
      "AT3G02680 atnbs1-1 Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions. 17672843\n",
      "\n",
      "AT3G02680 NBS1 190\n",
      "\n",
      "[{'last_insert_id()': 14}]\n",
      "14\n",
      "AT3G02850 CS3816 The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion. 17568770\n",
      "\n",
      "AT3G02850 SKOR 234\n",
      "\n",
      "[{'last_insert_id()': 15}]\n",
      "15\n",
      "AT3G02870 vtc4-1 ascorbate deficient 16595667\n",
      "\n",
      "AT3G02870 VTC4 311\n",
      "\n",
      "[{'last_insert_id()': 16}]\n",
      "16\n",
      "AT3G03260 hdg8-1 No visible phenotype. 16778018\n",
      "\n",
      "AT3G03260 HDG8 194\n",
      "\n",
      "[{'last_insert_id()': 17}]\n",
      "17\n",
      "AT4G14790 pdd17 Defective pollen development. 19237690\n",
      "\n",
      "AT4G14790 SUV3 312\n",
      "\n",
      "[{'last_insert_id()': 18}]\n",
      "18\n",
      "AT4G15210 bmy1-2 Plants cold-shocked for 6h have an increased starch content compared to wildtype. 16297066\n",
      "\n",
      "AT4G15210 BAM5 313\n",
      "\n",
      "[{'last_insert_id()': 19}]\n",
      "19\n",
      "AT4G15560 cla1-1 Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves. 10982425\n",
      "\n",
      "AT4G15560 DXS 219\n",
      "\n",
      "[{'last_insert_id()': 20}]\n",
      "20\n",
      "AT4G15570 maa3 Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT. 18772186\n",
      "\n",
      "AT4G15570 MAA3 294\n",
      "\n",
      "[{'last_insert_id()': 21}]\n",
      "21\n",
      "AT4G15802 Athspb-2 Early flowering, reduced fertility, aborted seeds. 20388662\n",
      "\n",
      "AT4G15802 HSBP 254\n",
      "\n",
      "[{'last_insert_id()': 22}]\n",
      "22\n",
      "AT4G15880 esd4-2 Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1. 17513499\n",
      "\n",
      "AT4G15880 ESD4 265\n",
      "\n",
      "[{'last_insert_id()': 23}]\n",
      "23\n",
      "AT4G16420 prz1-1 Altered response to auxin and cytokinin 12747832\n",
      "\n",
      "AT4G16420 ADA2B 279\n",
      "\n",
      "[{'last_insert_id()': 24}]\n",
      "24\n",
      "AT4G16480 atint4-2 No visible phenotype. 16603666\n",
      "\n",
      "AT4G16480 INT4 284\n",
      "\n",
      "[{'last_insert_id()': 25}]\n",
      "25\n",
      "AT5G10480 pas2-3 Segregates 25% embryo lethal. 18799749\n",
      "\n",
      "AT5G10480 PAS2 301\n",
      "\n",
      "[{'last_insert_id()': 26}]\n",
      "26\n",
      "AT5G10510 plt3-1 Short roots and shortened root meristem. 17960244\n",
      "\n",
      "AT5G10510 AIL6 310\n",
      "\n",
      "[{'last_insert_id()': 27}]\n",
      "27\n",
      "AT5G11110 kns2 Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile. 18779216\n",
      "\n",
      "AT5G11110 SPS2 232\n",
      "\n",
      "[{'last_insert_id()': 28}]\n",
      "28\n",
      "AT5G11260 hy5-101 Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller. 16891401\n",
      "\n",
      "AT5G11260 HY5 221\n",
      "\n",
      "[{'last_insert_id()': 29}]\n",
      "29\n",
      "AT5G11510 myb3r4-1 No visible phenotype. 17287251\n",
      "\n",
      "AT5G11510 MYB3R-4 336\n",
      "\n",
      "[{'last_insert_id()': 30}]\n",
      "30\n",
      "AT5G12200 pyd2-2 The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source. 19413687\n",
      "\n",
      "AT5G12200 PYD2 310\n",
      "\n",
      "[{'last_insert_id()': 31}]\n",
      "31\n",
      "AT5G13290 crn-1 Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3. 12345678\n",
      "\n",
      "AT5G13290 CRN 189\n",
      "\n",
      "[{'last_insert_id()': 32}]\n",
      "32\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "# Connect to the database\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='examanswer',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "file1 = open('./Germplasm.tsv', 'r')\n",
    "file2 = open('./LocusGene.tsv', 'r')\n",
    "\n",
    "#discard headers\n",
    "disc = file1.readline()\n",
    "disc = file2.readline()\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        # this is copy/paste from Problem #1\n",
    "        for line_file1 in file1.readlines():\n",
    "            line_file2 = file2.readline()\n",
    "            line_file1.strip()\n",
    "            line_file2.strip()\n",
    "            (loc,germ,phen,pm) = line_file1.split(\"\\t\") # split on tab\n",
    "            (loc2,gene,prot) = line_file2.split(\"\\t\") # split on tab\n",
    "\n",
    "            #print(loc,germ,phen,pm)\n",
    "            #print(loc2,gene,prot)\n",
    "            # Enter data to germplasm table\n",
    "            # take the last_insert_id\n",
    "            # enter data to the genes table using the germplasm_id\n",
    "            sql = \"insert into germplasm (locus, germplasm, phenotype, pubmed) values ('{}','{}','{}',{})\".format(loc,germ,phen,pm)\n",
    "            cursor.execute(sql)\n",
    "            \n",
    "            sql = \"select last_insert_id()\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "            germplasm_id = results[0]['last_insert_id()']\n",
    "            print(germplasm_id)\n",
    "\n",
    "            sql = \"insert into genes (germplasm_id, gene, protlength) values ({},'{}',{})\".format(germplasm_id, gene, prot)\n",
    "            cursor.execute(sql)\n",
    "        \n",
    "finally:\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sanity Check\n",
    "Look at the content of the tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>id</th>\n",
       "        <th>germplasm_id</th>\n",
       "        <th>gene</th>\n",
       "        <th>protlength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>1</td>\n",
       "        <td>1</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>2</td>\n",
       "        <td>2</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>3</td>\n",
       "        <td>3</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "        <td>4</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>5</td>\n",
       "        <td>5</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>6</td>\n",
       "        <td>6</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>7</td>\n",
       "        <td>7</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>8</td>\n",
       "        <td>8</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>9</td>\n",
       "        <td>9</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>10</td>\n",
       "        <td>10</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>11</td>\n",
       "        <td>11</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>12</td>\n",
       "        <td>12</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>13</td>\n",
       "        <td>13</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>14</td>\n",
       "        <td>14</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>15</td>\n",
       "        <td>15</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>16</td>\n",
       "        <td>16</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>17</td>\n",
       "        <td>17</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>18</td>\n",
       "        <td>18</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>19</td>\n",
       "        <td>19</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>20</td>\n",
       "        <td>20</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>21</td>\n",
       "        <td>21</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>22</td>\n",
       "        <td>22</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>23</td>\n",
       "        <td>23</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>24</td>\n",
       "        <td>24</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>25</td>\n",
       "        <td>25</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>26</td>\n",
       "        <td>26</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>27</td>\n",
       "        <td>27</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>28</td>\n",
       "        <td>28</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>29</td>\n",
       "        <td>29</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>30</td>\n",
       "        <td>30</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>31</td>\n",
       "        <td>31</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>32</td>\n",
       "        <td>32</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(1, 1, 'DCL1', 332),\n",
       " (2, 2, 'LHY', 290),\n",
       " (3, 3, 'CIPK9', 223),\n",
       " (4, 4, 'FKGP', 190),\n",
       " (5, 5, 'MRH6', 189),\n",
       " (6, 6, 'GEK1', 196),\n",
       " (7, 7, 'XERICO', 256),\n",
       " (8, 8, 'POT1A', 221),\n",
       " (9, 9, 'RPK2', 284),\n",
       " (10, 10, 'TMAC2', 300),\n",
       " (11, 11, 'RGP1', 301),\n",
       " (12, 12, 'BIG', 279),\n",
       " (13, 13, 'SEP2', 175),\n",
       " (14, 14, 'NBS1', 190),\n",
       " (15, 15, 'SKOR', 234),\n",
       " (16, 16, 'VTC4', 311),\n",
       " (17, 17, 'HDG8', 194),\n",
       " (18, 18, 'SUV3', 312),\n",
       " (19, 19, 'BAM5', 313),\n",
       " (20, 20, 'DXS', 219),\n",
       " (21, 21, 'MAA3', 294),\n",
       " (22, 22, 'HSBP', 254),\n",
       " (23, 23, 'ESD4', 265),\n",
       " (24, 24, 'ADA2B', 279),\n",
       " (25, 25, 'INT4', 284),\n",
       " (26, 26, 'PAS2', 301),\n",
       " (27, 27, 'AIL6', 310),\n",
       " (28, 28, 'SPS2', 232),\n",
       " (29, 29, 'HY5', 221),\n",
       " (30, 30, 'MYB3R-4', 336),\n",
       " (31, 31, 'PYD2', 310),\n",
       " (32, 32, 'CRN', 189)]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use examanswer\n",
    "#%sql select * from germplasm\n",
    "%sql select * from genes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 4\n",
    "\n",
    "##  Full join\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='examanswer',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "\n",
    "\n",
    "joinfile = open(\"fulljoin.tsv\", \"w\")\n",
    "joinfile.write(\"Locus\\tGermplasm\\tPheno\\tPMID\\tGene\\tProtlen\\n\")\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        sql = \"select locus, germplasm, phenotype, pubmed, gene, protlength from germplasm,genes where germplasm.id = genes.germplasm_id\"\n",
    "        cursor.execute(sql)\n",
    "            \n",
    "        results = cursor.fetchall()\n",
    "        for result in results:\n",
    "            locus = result['locus']\n",
    "            germplasm = result['germplasm']\n",
    "            phenotype = result['phenotype']\n",
    "            gene = result['gene']\n",
    "            pubmed = result['pubmed']\n",
    "            protlength = result['protlength']\n",
    "            \n",
    "            joinfile.write(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(locus, germplasm, phenotype, pubmed, gene, protlength))\n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGermplasm\tPheno\tPMID\tGene\tProtlen\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\n"
     ]
    }
   ],
   "source": [
    "joinfile = open(\"fulljoin.tsv\", \"r\")\n",
    "print(joinfile.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Join only for SKOR and MAA3\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'gene': 'SKOR', 'protlength': 234}\n",
      "{'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'gene': 'MAA3', 'protlength': 294}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='examanswer',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "\n",
    "joinfile = open(\"fulljoin.tsv\", \"a\")   # open for append\n",
    "joinfile.write(\"\\n\\n\\n REPORT ONLY FOR SKOR and MAA3 \\n\\n\")\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        sql = \"select locus, germplasm, phenotype, pubmed, gene, protlength from germplasm,genes where germplasm.id = genes.germplasm_id AND (gene='SKOR' OR gene='MAA3')\"\n",
    "        cursor.execute(sql)\n",
    "            \n",
    "        results = cursor.fetchall()\n",
    "        for result in results:\n",
    "            locus = result['locus']\n",
    "            germplasm = result['germplasm']\n",
    "            phenotype = result['phenotype']\n",
    "            gene = result['gene']\n",
    "            pubmed = result['pubmed']\n",
    "            protlength = result['protlength']\n",
    "            print(result)\n",
    "            joinfile.write(\"{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(locus, germplasm, phenotype, pubmed, gene, protlength))\n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "\n",
    "joinfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGermplasm\tPheno\tPMID\tGene\tProtlen\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\n",
      "\n",
      "\n",
      " REPORT ONLY FOR SKOR and MAA3 \n",
      "\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "\n"
     ]
    }
   ],
   "source": [
    "joinfile = open(\"fulljoin.tsv\", \"r\")\n",
    "print(joinfile.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='examanswer',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "\n",
    "joinfile = open(\"fulljoin.tsv\", \"a\")   # open for append\n",
    "joinfile.write(\"\\n\\n\\n REPORT COUNT PER CHROMOSOME \\n\\n\")\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for chr in [1,2,3,4,5]:\n",
    "            sql = \"select COUNT(*) from germplasm where locus like 'AT{}G%'\".format(chr)\n",
    "            cursor.execute(sql)    \n",
    "            results = cursor.fetchall()\n",
    "            count = results[0]['COUNT(*)']\n",
    "            joinfile.write(\"there are {} loci on Chromosoe {}\\n\".format(count, chr))\n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "\n",
    "joinfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGermplasm\tPheno\tPMID\tGene\tProtlen\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\n",
      "\n",
      "\n",
      " REPORT ONLY FOR SKOR and MAA3 \n",
      "\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "\n",
      "\n",
      "\n",
      " REPORT COUNT PER CHROMOSOME \n",
      "\n",
      "there are 4 loci on Chromosoe 1\n",
      "there are 4 loci on Chromosoe 2\n",
      "there are 9 loci on Chromosoe 3\n",
      "there are 8 loci on Chromosoe 4\n",
      "there are 7 loci on Chromosoe 5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "joinfile = open(\"fulljoin.tsv\", \"r\")\n",
    "print(joinfile.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='examanswer',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor,\n",
    "                             autocommit=True)\n",
    "\n",
    "\n",
    "joinfile = open(\"fulljoin.tsv\", \"a\")   # open for append\n",
    "joinfile.write(\"\\n\\n\\n REPORT AVERAGE LENGTH PER CHROMOSOME \\n\\n\")\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        for chr in [1,2,3,4,5]:\n",
    "            sql = \"select AVG(protlength) from genes, germplasm where germplasm.id = genes.germplasm_id and locus like 'AT{}G%'\".format(chr)\n",
    "            cursor.execute(sql)    \n",
    "            results = cursor.fetchall()\n",
    "            count = results[0]['AVG(protlength)']\n",
    "            joinfile.write(\"Averagte protein length of {} Chromosoe {}\\n\".format(count, chr))\n",
    "        \n",
    "finally:\n",
    "    connection.close()\n",
    "\n",
    "joinfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Locus\tGermplasm\tPheno\tPMID\tGene\tProtlen\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\n",
      "\n",
      "\n",
      " REPORT ONLY FOR SKOR and MAA3 \n",
      "\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "\n",
      "\n",
      "\n",
      " REPORT COUNT PER CHROMOSOME \n",
      "\n",
      "there are 4 loci on Chromosoe 1\n",
      "there are 4 loci on Chromosoe 2\n",
      "there are 9 loci on Chromosoe 3\n",
      "there are 8 loci on Chromosoe 4\n",
      "there are 7 loci on Chromosoe 5\n",
      "\n",
      "\n",
      "\n",
      " REPORT AVERAGE LENGTH PER CHROMOSOME \n",
      "\n",
      "Averagte protein length of 258.7500 Chromosoe 1\n",
      "Averagte protein length of 215.5000 Chromosoe 2\n",
      "Averagte protein length of 252.0000 Chromosoe 3\n",
      "Averagte protein length of 277.5000 Chromosoe 4\n",
      "Averagte protein length of 271.2857 Chromosoe 5\n",
      "\n"
     ]
    }
   ],
   "source": [
    "joinfile = open(\"fulljoin.tsv\", \"r\")\n",
    "print(joinfile.read())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
